lib/rules/no-undef.js

```diff
@@ -110,6 +113,7 @@ module.exports = function(context) {
             variable = new escope.Variable(node.local.name, scope);
         variable.defs.push(node);
```

Comment:
With the change you just suggested and the test from this PR:

```
1) eslint verify() should not crash due to no-undef mutating escope data:
     TypeError: Cannot read property 'loc' of undefined
      at EventEmitter.module.exports.api.report (~/code/eslint/eslint/lib/eslint.js:782:28)
      at RuleContext.report (~/code/eslint/eslint/lib/rule-context.js:98:16)
      at EventEmitter.Program:exit (~/code/eslint/eslint/lib/rules/no-unused-vars.js:136:29)
      at EventEmitter.emit (events.js:129:20)
      at Controller.controller.traverse.leave (~/code/eslint/eslint/lib/eslint.js:743:25)
      at Controller.__execute (~/code/eslint/eslint/node_modules/estraverse/estraverse.js:393:31)
      at Controller.traverse (~/code/eslint/eslint/node_modules/estraverse/estraverse.js:481:28)
      at EventEmitter.module.exports.api.verify (~/code/eslint/eslint/lib/eslint.js:728:24)
      at Context.<anonymous> (~/code/eslint/eslint/tests/lib/eslint.js:2315:20)
      at callFn (~/code/eslint/eslint/node_modules/mocha/lib/runnable.js:266:21)
      at Test.Runnable.run (~/code/eslint/eslint/node_modules/mocha/lib/runnable.js:259:7)
      at Runner.runTest (~/code/eslint/eslint/node_modules/mocha/lib/runner.js:387:10)
      at ~/code/eslint/eslint/node_modules/mocha/lib/runner.js:470:12
      at next (~/code/eslint/eslint/node_modules/mocha/lib/runner.js:312:14)
      at ~/code/eslint/eslint/node_modules/mocha/lib/runner.js:322:7
      at next (~/code/eslint/eslint/node_modules/mocha/lib/runner.js:257:23)
      at Immediate._onImmediate (~/code/eslint/eslint/node_modules/mocha/lib/runner.js:289:5)
      at processImmediate [as _immediateCallback] (timers.js:358:17)
```

I fear this is the same kind of rabbit hole babel-eslint gets into, trying to monkey patch until dependencies get updated.